### PR TITLE
XD-3482: Add jdbc sink

### PR DESCRIPTION
**Pull request by ericbottard**
The module is, I believe, a "smarter" rewrite of the XD jdbc sink:
- Allows single inserts or opt-in batching support (the payload needs to be convertible to a `Collection`)
- Uses SpEL for computing the column value, using the `columns` option, which is a mapping from column name to SpEL. 
- A dedicated converter allows a shorthand configuration for the above though, where `columns=a,b` is equivalent to `columns=a:a, b:b`
- The SpEL is to be evaluated against both the whole message and `payload` in the single case, and collection elements in the batching case.
- This still allows XD1 compatible behavior, such as inserting the whole payload (the default is indeed `payload: payload.toString()`)
- Leverages Boot database support, so props like `spring.datasource.url` are to be used.

Comes with no driver OOTB for now, but can easily be added with `--includes`. For example using dataflow:

```
stream create foo --definition "http | jdbc --includes=com.h2database:h2:1.4.189 --spring_h2_console_enabled=true --initialize=true" --deploy
```

and go see the console at `/h2-console`

### CODE REVIEW

**Comment by trisberg** [comment]
I wonder if we don't also need `url`, `username`, `password` properties so we can override `spring.datasource` values. This could come in handy when we use this module together with a jdbc sink reading from one db and writing to a different one - e.g. `"jdbc | jdbc --url=jdbc:oracle:thin:@//localhost:1521/orcl --username=scott --password=tiger --includes=com.oracle:jdbc:12.1.0.2"` where the first module would read from the "default" `spring.datasource`.

**Comment by ericbottard** [comment]
There should not be a "default" datasource IMHO. All modules should be isolated from each other and the runner

**Comment by trisberg** [comment]
If we use a config server there could be one, plus ATM I don't think that SCDF pass along any config setting to the modules other than the explicitly specified ones.

**Comment by ericbottard** [comment]
> If we use a config server there could be one, 

I don't think so. If we used a config server, there would be defaults for _e.g._ `spring.datasource.url` (which would apply to the first module) and which the second module would override.

> plus ATM I don't think that SCDF pass along any config setting to the modules other than the explicitly specified ones.

correct

**Comment by ericbottard** [comment]
If anyone is wondering about why underscores in the example above, this is why: https://github.com/spring-cloud/spring-cloud-dataflow/pull/96

**Comment by trisberg** [comment]
Also, should we add a `types` option where we could specify the SQL data types for the columns, like `id:INTEGER, name:VARCHAR, created:TIMESTAMP` instead of always using VARCHAR(2000)

**Comment by ericbottard** [comment]
> Also, should we add a types option where we could specify the SQL data types for the columns, like id:INTEGER, name:VARCHAR, created:TIMESTAMP instead of always using VARCHAR(2000)

yes, thought about that (the XD1 behavior is really limiting) but wondered about the best notation

Could also be
`--columns=id(INTEGER):foo.toString(), created(TIMESTAMP)`

(I already introduced a custom converter anyway)

**Comment by trisberg** [comment]
That works, maybe use `[]` to allow including the length like:

`--columns=id[INTEGER]:foo.toString(), name[VARCHAR(200)], created[TIMESTAMP]`

**Comment by ericbottard** [comment]
Shall we try to merge this first, or do you want me to amend this PR?

**Comment by trisberg** [comment]
Whatever is easier, I'll test what we have now and then we can decide how we can best add this new functionality.

**Comment by ericbottard** [comment]
> why use spring-integration-jdbc? wouldn't spring-boot-starter-jdbc provide more features...

My bad, I did not understand what you meant at first. Did not realize I left an SI dependency there, even though I did switch to a pure jdbc approach.

I also see what you mean wrt auto-configuration now: for any non-embedded db to work, we need a pool implementatio, which spring-boot-starter-jdbc provides.

Doing some more tests and will amend the PR.

**Comment by ericbottard** [comment]
@trisberg OK, I updated the code to ease auto-configuration of a non-embedded bd.
Successfully tested in the non-ModuleLauncher case, adding for example hsqldb to the pom (scope=runtime) and doing

```
java -jar jdbc-sink/target/jdbc-sink-1.0.0.BUILD-SNAPSHOT-exec.jar --spring.datasource.url=jdbc:hsqldb:hsql://localhost:9101/xdjob --spring.cloud.stream.bindings.input=application --initialize=true --columns=a:payload.charAt\(0\).toString\(\),b:payload.length\(\),c:payload.charAt\(payload.length\(\)-1\).toString\(\)
```

against the db we ship with XD.

Still need to decide which drivers we want/can add to the pom, if any. (they'd get embedded in the `-exec` jar, so need to consider licensing issues)

**Comment by ericbottard** [comment]
Made some changes to make sure the correct `EvaluationContext` will be used once [Mark's PR](https://github.com/spring-cloud/spring-cloud-stream/compare/master...markfisher:XD-3581) is merged.

Successfully tested the following:

```
stream create foo --definition "http --server.port=8082| jdbc --includes=com.h2database:h2:1.4.189 --spring_h2_console_enabled=true --initialize=true --inputType=application/json --server.port=8081 --columns=a:payload.key.toString()" --deploy

curl -v -d '{"key": 5}' http://localhost:8082/messages -H 'Content-Type: text/plain'
```

```
stream create foo --definition "http --server.port=8082| jdbc --includes=com.h2database:h2:1.4.189 --spring_h2_console_enabled=true --initialize=true --inputType=application/x-spring-tuple --server.port=8081 --columns=a:payload.key.toString()" --deploy

curl -v -d '{"key": 5}' http://localhost:8082/messages -H 'Content-Type: application/json'
```

with @markfisher 's changes applied

File: jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSink.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	protected JdbcSink(JdbcOperations jdbcOperations, String tableName, Set<String> columns) {
+		this.jdbcOperations = jdbcOperations;
+		this.tableName = tableName;
+		this.sql = generateSql(columns);
+	}
+
+	@Autowired
+	public void setBeanFactory(BeanFactory beanFactory) {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(beanFactory);
+	}
```
Comment (trisberg) [inline jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/JdbcSink.java:53]: This should be done in `afterPropertiesSet()` according to Artem --
Artem Bilan [1:44 PM] 
No, you can't do that on the `setBeanFactory()`. Move that injection to the `afterPropertiesSet()`

File: jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/DefaultInitializationScriptResource.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		result.append("CREATE TABLE ").append(properties.getTableName()).append('(');
+		int i = 0;
+		for (String column : properties.getColumns().keySet()) {
+			if (i++ > 0) {
+				result.append(", ");
+			}
+			result.append(column).append(" VARCHAR(2000)");
+		}
+		result.append(");\n");
+		System.out.println(result);
```
Comment (markfisher) [inline jdbc-sink/src/main/java/org/springframework/cloud/stream/module/jdbc/DefaultInitializationScriptResource.java]: this system.out should be removed (or replaced with logging)

**Comment by ericbottard** [comment]
Addressed review comments, rebased

**Comment by trisberg** [comment]
I can get it to work if I let the module create the table. If I create the table ahead of time and use `--inititialize=false` I get some errors trying to insert.

My table def (MySQL):

```
CREATE TABLE
    testdata
    (
        id BIGINT,
        name VARCHAR(2000),
        age SMALLINT
    )
    ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

Module startup command:

```
$ java -jar spring-cloud-stream-module-launcher/target/spring-cloud-stream-module-launcher-1.0.0.BUILD-SNAPSHOT.jar --modules=org.springframework.cloud.stream.module:jdbc-sink:1.0.0.BUILD-SNAPSHOT --args.0.includes=mysql:mysql-connector-java:5.1.36 --args.0.server.port=8081 --args.0.spring.cloud.stream.bindings.input=data --args.0.spring.datasource.url=jdbc:mysql://localhost:3306/test --args.0.spring.datasource.username=spring --args.0.spring.datasource.password=***** --args.0.spring.datasource.driverClassName=com.mysql.jdbc.Driver --args.0.tableName=testdata --args.0.columns='id,name,age' --args.0.initialize=false
```

Test data posted via http-source:

```
curl -H 'Content-Type: application/json' -X POST --data '{"id": 12, "name": "Sven", "age": 44}' http://localhost:8080/messages
```

The exception:

```
2015-10-19 16:32:13.886  INFO 40892 --- [           main] o.s.c.s.module.jdbc.JdbcSinkApplication  : Started JdbcSinkApplication in 5.133 seconds (JVM running for 9.132)
2015-10-19 16:32:26.390  INFO 40892 --- [hannel-adapter1] o.s.b.f.xml.XmlBeanDefinitionReader      : Loading XML bean definitions from class path resource [org/springframework/jdbc/support/sql-error-codes.xml]
2015-10-19 16:32:26.431  INFO 40892 --- [hannel-adapter1] o.s.jdbc.support.SQLErrorCodesFactory    : SQLErrorCodes loaded: [DB2, Derby, H2, HSQL, Informix, MS-SQL, MySQL, Oracle, PostgreSQL, Sybase, Hana]
2015-10-19 16:32:29.450 ERROR 40892 --- [hannel-adapter1] o.s.c.s.b.r.RedisMessageChannelBinder$1  : Failed to deliver message; retries exhausted; message sent to queue 'ERRORS:data' 
java.io.NotSerializableException: org.springframework.integration.json.JsonPropertyAccessor$ToStringFriendlyJsonNode
    at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1183) ~[na:1.7.0_67]
    at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347) ~[na:1.7.0_67]
    at com.mysql.jdbc.PreparedStatement.setSerializableObject(PreparedStatement.java:3876) ~[mysql-connector-java-5.1.36.jar!/:na]
    ... 62 common frames omitted
Wrapped by: java.sql.SQLException: Invalid argument value: java.io.NotSerializableException
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:998) ~[mysql-connector-java-5.1.36.jar!/:na]
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:937) ~[mysql-connector-java-5.1.36.jar!/:na]
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:926) ~[mysql-connector-java-5.1.36.jar!/:na]
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:872) ~[mysql-connector-java-5.1.36.jar!/:na]
    at com.mysql.jdbc.PreparedStatement.setSerializableObject(PreparedStatement.java:3887) ~[mysql-connector-java-5.1.36.jar!/:na]
    at com.mysql.jdbc.PreparedStatement.setObject(PreparedStatement.java:3603) ~[mysql-connector-java-5.1.36.jar!/:na]
    at org.springframework.cloud.stream.module.jdbc.SimpleJdbcSink$1.setValues(SimpleJdbcSink.java:66) ~[jdbc-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:na]
    at org.springframework.jdbc.core.JdbcTemplate$2.doInPreparedStatement(JdbcTemplate.java:871) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.core.JdbcTemplate$2.doInPreparedStatement(JdbcTemplate.java:866) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:629) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    ... 57 common frames omitted
Wrapped by: org.springframework.dao.TransientDataAccessResourceException: PreparedStatementCallback; SQL [INSERT INTO testdata(id, name, age) VALUES (?, ?, ?)]; Invalid argument value: java.io.NotSerializableException; nested exception is java.sql.SQLException: Invalid argument value: java.io.NotSerializableException
    at org.springframework.jdbc.support.SQLStateSQLExceptionTranslator.doTranslate(SQLStateSQLExceptionTranslator.java:108) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:73) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:645) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:866) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:927) ~[spring-jdbc-4.2.1.RELEASE.jar!/:na]
    at org.springframework.cloud.stream.module.jdbc.SimpleJdbcSink.handle(SimpleJdbcSink.java:50) ~[jdbc-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:na]
    at org.springframework.cloud.stream.module.jdbc.JdbcSinkConfiguration.handleMessage(JdbcSinkConfiguration.java:73) ~[jdbc-sink-1.0.0.BUILD-SNAPSHOT-exec.jar!/:na]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_67]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_67]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_67]
    at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_67]
    at org.springframework.expression.spel.support.ReflectiveMethodExecutor.execute(ReflectiveMethodExecutor.java:112) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.ast.MethodReference.getValueInternal(MethodReference.java:102) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.ast.MethodReference.access$000(MethodReference.java:49) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.ast.MethodReference$MethodValueRef.getValue(MethodReference.java:342) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.ast.CompoundExpression.getValueInternal(CompoundExpression.java:88) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.ast.SpelNodeImpl.getTypedValue(SpelNodeImpl.java:131) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:330) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.integration.util.AbstractExpressionEvaluator.evaluateExpression(AbstractExpressionEvaluator.java:166) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.util.MessagingMethodInvokerHelper.processInternal(MessagingMethodInvokerHelper.java:317) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.util.MessagingMethodInvokerHelper.process(MessagingMethodInvokerHelper.java:155) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:93) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    ... 38 common frames omitted
Wrapped by: org.springframework.messaging.MessageHandlingException: ; nested exception is org.springframework.dao.TransientDataAccessResourceException: PreparedStatementCallback; SQL [INSERT INTO testdata(id, name, age) VALUES (?, ?, ?)]; Invalid argument value: java.io.NotSerializableException; nested exception is java.sql.SQLException: Invalid argument value: java.io.NotSerializableException
    at org.springframework.integration.handler.MethodInvokingMessageProcessor.processMessage(MethodInvokingMessageProcessor.java:96) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:89) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:99) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:147) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:120) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutput(AbstractMessageProducingHandler.java:231) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractMessageProducingHandler.produceOutput(AbstractMessageProducingHandler.java:154) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractMessageProducingHandler.sendOutputs(AbstractMessageProducingHandler.java:102) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:105) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:147) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:120) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$1.doWithRetry(RedisMessageChannelBinder.java:264) ~[spring-cloud-stream-binder-redis-1.0.0.BUILD-SNAPSHOT.jar!/:na]
    at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$1.doWithRetry(RedisMessageChannelBinder.java:260) ~[spring-cloud-stream-binder-redis-1.0.0.BUILD-SNAPSHOT.jar!/:na]
    at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:263) [spring-retry-1.1.2.RELEASE.jar!/:na]
    at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:168) [spring-retry-1.1.2.RELEASE.jar!/:na]
    at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1.doSend(RedisMessageChannelBinder.java:260) [spring-cloud-stream-binder-redis-1.0.0.BUILD-SNAPSHOT.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:105) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.popMessageAndSend(RedisQueueMessageDrivenEndpoint.java:224) [spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.access$300(RedisQueueMessageDrivenEndpoint.java:54) [spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint$ListenerTask.run(RedisQueueMessageDrivenEndpoint.java:323) [spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.util.ErrorHandlingTaskExecutor$1.run(ErrorHandlingTaskExecutor.java:55) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at java.lang.Thread.run(Thread.java:745) [na:1.7.0_67]
2015-10-19 16:32:29.452 ERROR 40892 --- [hannel-adapter1] o.s.c.s.b.r.RedisMessageChannelBinder$1  : Failed to deliver message
java.lang.IllegalArgumentException: EvaluationContext must not be null
    at org.springframework.util.Assert.notNull(Assert.java:115) ~[spring-core-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.ExpressionState.<init>(ExpressionState.java:85) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:329) ~[spring-expression-4.2.1.RELEASE.jar!/:na]
    at org.springframework.integration.redis.outbound.RedisQueueOutboundChannelAdapter.handleMessageInternal(RedisQueueOutboundChannelAdapter.java:115) ~[spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    ... 17 common frames omitted
Wrapped by: org.springframework.messaging.MessageHandlingException: error occurred in message handler [org.springframework.integration.redis.outbound.RedisQueueOutboundChannelAdapter@2693ec5f]; nested exception is java.lang.IllegalArgumentException: EvaluationContext must not be null
    at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:139) ~[spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$2.recover(RedisMessageChannelBinder.java:277) ~[spring-cloud-stream-binder-redis-1.0.0.BUILD-SNAPSHOT.jar!/:na]
    at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1$2.recover(RedisMessageChannelBinder.java:267) ~[spring-cloud-stream-binder-redis-1.0.0.BUILD-SNAPSHOT.jar!/:na]
    at org.springframework.retry.support.RetryTemplate.handleRetryExhausted(RetryTemplate.java:458) ~[spring-retry-1.1.2.RELEASE.jar!/:na]
    at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:320) ~[spring-retry-1.1.2.RELEASE.jar!/:na]
    at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:168) ~[spring-retry-1.1.2.RELEASE.jar!/:na]
    at org.springframework.cloud.stream.binder.redis.RedisMessageChannelBinder$1.doSend(RedisMessageChannelBinder.java:260) ~[spring-cloud-stream-binder-redis-1.0.0.BUILD-SNAPSHOT.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:442) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:392) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105) [spring-messaging-4.2.1.RELEASE.jar!/:na]
    at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:105) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.popMessageAndSend(RedisQueueMessageDrivenEndpoint.java:224) [spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.access$300(RedisQueueMessageDrivenEndpoint.java:54) [spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint$ListenerTask.run(RedisQueueMessageDrivenEndpoint.java:323) [spring-integration-redis-4.2.0.RELEASE.jar!/:na]
    at org.springframework.integration.util.ErrorHandlingTaskExecutor$1.run(ErrorHandlingTaskExecutor.java:55) [spring-integration-core-4.2.0.RELEASE.jar!/:na]
    at java.lang.Thread.run(Thread.java:745) [na:1.7.0_67]
2015-10-19 16:32:29.454 ERROR 40892 --- [hannel-adapter1] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessageDeliveryException: failed to send message to channel 'data.bridge' within timeout: -1
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:118)
    at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45)
    at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105)
    at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:105)
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.popMessageAndSend(RedisQueueMessageDrivenEndpoint.java:224)
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint.access$300(RedisQueueMessageDrivenEndpoint.java:54)
    at org.springframework.integration.redis.inbound.RedisQueueMessageDrivenEndpoint$ListenerTask.run(RedisQueueMessageDrivenEndpoint.java:323)
    at org.springframework.integration.util.ErrorHandlingTaskExecutor$1.run(ErrorHandlingTaskExecutor.java:55)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by trisberg** [comment]
I think one problem is that we are getting values of type `org.springframework.integration.json.JsonPropertyAccessor$ToStringFriendlyJsonNode` which probably cause this error

**Comment by ericbottard** [comment]
> I think one problem is that we are getting values of type org.springframework.integration.json.JsonPropertyAccessor$ToStringFriendlyJsonNode which probably cause this error

Yes, you have to use `.toString()` in the SpEL expression. It's one of my rants about our type conversion support for JSON, but it's not new (and should be addressed globally, if at all possible)

PS: I rebased and re-pushed the code

**Comment by trisberg** [comment]
Confused, I don't see any new changes.

**Comment by trisberg** [comment]
If I leave out a property I'd assume to get a NULL value in that column, right now we seem to require that all properties are present and I get an exception.

**Comment by markfisher** [comment]
@ericbottard wrt your rant about conversion, this might be of interest: https://github.com/spring-cloud/spring-cloud-stream/issues/156

**Comment by ericbottard** [comment]
> Confused, I don't see any new changes.

Yes, what I was saying is that it's up to the user to use explicit SpEL with an explicit `.toString()` appended to what returns an `JsonPropertyAccessor$ToStringFriendlyJsonNode`

**Comment by ericbottard** [comment]
> If I leave out a property I'd assume to get a NULL value in that column, right now we seem to require that all properties are present and I get an exception.

Not sure what you mean. The update statement is crafted from the list of properties that are provided. If you leave out a property (I assume you meant "from that list") then it is untouched (why would you want `NULL` rather than something else. `NULL` is a value like any other IMO). If the user wants a `NULL`, he'd have to make it explicit.
Moreover, I guess this is consistent with what we had in XD, no?

**Comment by trisberg** [comment]
I think XD allows missing properties and inserts a NULL in that case. If I declare a stream using `--columns='id,name,age'` and pass in `{"id": 12, "name": "Bob"}` I'd expect that to work and the age column to be NULL for that row.

**Comment by ericbottard** [comment]
Ah, gotcha, misunderstood what you were saying (I thought you meant `--columns='id,name,age'` but the table actually had a fourth column)

Nevertheless, is this really what we want? `--columns=a,b` actually is a shorthand for `--columns=a:a,b:b` so `b` needs to be there, somehow (or have the PropertyAccessor for the underlying message decide to return `null` for missing values).

Maybe we should simply change (and document) the behavior of the shorthand? Or would you say that when you explicitly defined `--columns=a:a,b:foo` and either `a` or `foo` don't exist you'd expect to have `null` silently inserted? I think this would be wrong

**Comment by trisberg** [comment]
The current XD implementation supports missing properties and inserts a NULL value. That to me seems to be the "least surprise" option. I would think that it is a common occurrence where not all messages contain the exact set of fields. If we are going to require every field to be present in every message we would need to introduce a way to handle a `null` value. Right now I get `java.sql.SQLException: Incorrect integer value: 'null' for column 'age' at row 1`

**Comment by trisberg** [comment]
Replaced with new PR https://github.com/spring-cloud/spring-cloud-stream-modules/pull/55

**Pull request closed without merging**